src/Components/Web.JS/src/Services/RootComponentManager.ts

```diff
@@ -30,39 +32,58 @@ type RootComponentInfo = {
   interactiveComponentId?: number;
 }
 
-let resolveAutoMode: () => 'server' | 'webassembly' | null = () => {
-  throw new Error('No auto mode resolver has been attached');
-};
-
-export function attachAutoModeResolver(resolver: () => 'server' | 'webassembly' | null) {
-  resolveAutoMode = resolver;
-}
-
-export class RootComponentManager {
+export class RootComponentManager implements NavigationEnhancementCallbacks {
   private readonly _activeDescriptors = new Set<ComponentDescriptor>();
 
   private readonly _descriptorsPendingInteractivityById: { [id: number]: ComponentDescriptor } = {};
 
   private readonly _rootComponentInfoByDescriptor = new Map<ComponentDescriptor, RootComponentInfo>();
 
+  private _didWebAssemblyFailToLoadQuickly = false;
+
+  public constructor() {
+    this.updateAllRootComponentsAfterRendererAttached(WebRendererId.Server);
+    this.updateAllRootComponentsAfterRendererAttached(WebRendererId.WebAssembly);
```

Comment:
Each of these calls will wait for `rendererAttached` and then call `updateAllRootComponents`.
So technically, if there are both webassembly and server interactive root components, all will flicker twice? I am just wondering, if it will make sense to wait a little bit longer and call the `updateRootComponents` only once?

Reply:
No flickering happens here. What these calls do is wait for the specific renderer to attach (which indicates that the platform is fully initialized), then call `updateAllRootComponents()`. The `updateAllRootComponents()` function looks at each root component to determine if:
* It's never been seen before, in which case it gets enabled for interactivity
* Or, it's been seen before, but it's parameters have changed, in which case send the updated parameters to .NET
* Or, it's been seen before, but now it's been removed from the page in an SSR update, so we dispose it in .NET
* Or, none of the above cases happen, in which case it's a no-op for the given component

It can kind of be thought of as calling `StateHasChanged()`, except for that it doesn't always trigger a re-render. There isn't really a downside to calling it in cases where we don't end up needing to.